File: document/src/libs/algorithms/fast_rrt.ts
import {GridMap} from "../grid";
import {TraceEvent} from "../trace/types";
import {NumpyRandom} from "./numpy_rng";
import {Point, SamplingGrid} from "./sampling_space";
import {pathLength, Tree} from "./rrt";
import {insertBestParent, nearRadius, rewireStep} from "./rrt_star";

// 브라우저 라이브 데모용 Fast-RRT (Wu et al. 2021, Applied Sciences 11(24):11777).
// 저장소 구현을 그대로 미러한다: RRT* 트리 위에 세 가지 가속을 얹는다 —
// Fast-Sampling(기존 노드 근방 표본 거부로 미탐사 공간 집중), Random Steering(직선
// 확장이 막히면 무작위 방향으로 재시도), Fast-Optimal(경로가 생기면 삼각 부등식으로
// 지름길 pruning). planner RNG와 map RNG가 같은 seed의 독립 numpy 스트림이라는 것,
// 그리고 goal_bias 동전과 random steering 각도는 planner RNG, 표본은 map RNG라는
// draw 순서까지 미러하므로 같은 seed면 python demo와 표본·트리·개선 열이 일치한다.
export interface FastRRTOptions {
    map: GridMap;
    start: Point;
    goal: Point;
    maxIterations: number;
    stepSize: number;
    goalBias: number;
    goalTolerance: number;
    neighborRadius: number;
    radiusMode: "fixed" | "shrinking";
    rggGamma: number;
    reachedRadius: number;
    steeringAttempts: number;
    seed: number;
}

// Fast-Sampling (Wu et al. 2021): goal-biased 표본은 goal 도달을 위해 항상 받고,
// 자유 표본은 어떤 트리 노드의 reached_radius 안이면 (steering_attempts 상한 내)
// 거부해 미탐사 공간에 집중시킨다. 이로써 search-time variance가 줄어든다.
function fastSample(
    space: SamplingGrid, tree: Tree, goal: Point, goalBias: number,
    reachedRadius: number, steeringAttempts: number, rng: NumpyRandom,
): Point {
    if (rng.random() < goalBias) return goal
    let q = space.sample()
    for (let a = 0; a < steeringAttempts; a++) {
        if (tree.points.every((node) => space.distance(q, node) > reachedRadius)) break
        q = space.sample()
    }
    return q
}

// Random Steering (Wu et al. 2021): 직선 확장이 막히면 무작위 방향으로 한 스텝씩
// 시도해 첫 충돌 없는 스텝을 취한다 (좁은 통로 통과에 도움).
function randomSteer(
    space: SamplingGrid, qNear: Point, qRand: Point, stepSize: number,
    steeringAttempts: number, rng: NumpyRandom,
): Point | null {
    const qNew = space.steer(qNear, qRand, stepSize)
    if (space.isMotionValid(qNear, qNew)) return qNew
    for (let a = 0; a < steeringAttempts; a++) {
        const angle = rng.uniform(0, 2 * Math.PI)
        const cand: Point = [qNear[0] + stepSize * Math.cos(angle),
                             qNear[1] + stepSize * Math.sin(angle)]
        if (space.isMotionValid(qNear, cand)) return cand
    }
    return null
}

// Fast-Optimal (Wu et al. 2021): 유지된 각 waypoint에서 충돌 없는 직선 구간으로
// 닿는 가장 먼 이후 waypoint로 뛰는 greedy 지름길 pass. C++ shortcut_prune과
// i/j 스캔 순서·is_motion_valid 호출 열이 같아 양 언어가 동일 결과를 낸다.
function shortcut(space: SamplingGrid, path: Point[]): Point[] {
    if (path.length <= 2) return path
    const out: Point[] = [path[0]]
    let i = 0
    while (i < path.length - 1) {
        let j = path.length - 1
        while (j > i + 1 && !space.isMotionValid(path[i], path[j])) j--
        out.push(path[j])
        i = j
    }
    return out
}

export function runFastRRT(opts: FastRRTOptions): TraceEvent[] {
    const {map, start, goal, maxIterations, stepSize, goalBias, goalTolerance,
           neighborRadius, radiusMode, rggGamma, reachedRadius, steeringAttempts, seed} = opts
    const space = new SamplingGrid(map, seed)
    const rng = new NumpyRandom(seed)
    const events: TraceEvent[] = []
    let seq = 0
    const emit = (ev: Omit<TraceEvent, "seq">) => events.push({seq: seq++, ...ev})
    emit({
        event: "planning_started",
        algorithm: "fast_rrt",
        params: {max_iterations: maxIterations, step_size: stepSize, goal_bias: goalBias,
                 goal_tolerance: goalTolerance, neighbor_radius: neighborRadius,
                 radius_mode: radiusMode, rgg_gamma: rggGamma, reached_radius: reachedRadius,
                 steering_attempts: steeringAttempts, seed},
    })

    const tree = new Tree(start)
    // goal은 트리에 넣지 않는다 (성장/rewire 후보가 되면 안 된다). 지름길 pruning된
    // 최선 경로와 그 비용만 추적한다 (anytime).
    let bestPath: Point[] = []
    let bestCost = Infinity
    let iterations = 0
    for (let it = 0; it < maxIterations; it++) {
        iterations++
        const qRand = fastSample(space, tree, goal, goalBias, reachedRadius, steeringAttempts, rng)
        emit({event: "sample_drawn", state: [qRand[0], qRand[1]]})
        const nearIdx = tree.nearest(qRand, space)
        const qNear = tree.points[nearIdx]
        const qNew = randomSteer(space, qNear, qRand, stepSize, steeringAttempts, rng)
        if (qNew === null) continue

        const radius = nearRadius(radiusMode, neighborRadius, rggGamma, tree.size)
        const neighborhood = tree.near(qNew, radius, space)
        const newIdx = insertBestParent(space, tree, qNew, nearIdx, neighborhood, emit)
        rewireStep(space, tree, newIdx, qNew, neighborhood, emit)

        if (space.distance(qNew, goal) <= goalTolerance && space.isMotionValid(qNew, goal)) {
            const raw = [...tree.pathTo(newIdx), goal]
            const pruned = shortcut(space, raw)
            const prunedCost = pathLength(space, pruned)
            if (prunedCost < bestCost) {
                bestCost = prunedCost
                bestPath = pruned
                emit({event: "path_found", path: pruned.map((p) => [p[0], p[1]]), cost: prunedCost})
            }
        }
    }

    const success = bestPath.length > 0
    const cost = success ? bestCost : 0
    emit({
        event: "planning_finished",
        success,
        metrics: {path_cost: cost, expanded_nodes: tree.size - 1,
                  samples: iterations, tree_size: tree.size, iterations},
    })
    return events
}

File: python/navigation/global_planning/sampling/fast_rrt.py
"""Fast-RRT (Wu et al. 2021, Applied Sciences 11(24):11777).

Improved-RRT + Fast-Optimal on top of an RRT*-style tree:
  * Fast-Sampling: reject a random sample within `reached_radius` of any existing
    tree node, so sampling concentrates on unreached space (lower search-time
    variance).
  * Random Steering: when the straight extension toward the sample is blocked,
    retry up to `steering_attempts` random directions (one step_size each) and
    take the first collision-free step — helps through narrow passages.
  * Fast-Optimal: once a feasible path exists, shortcut-prune it via the triangle
    inequality (drop waypoints whose bypass segment is collision-free) and keep
    the min-cost path; anytime, like RRT*.

Divergence from the paper (deliberate): Wu et al.'s framework re-initialises a
PLAIN RRT (no rewiring) per outer iteration and their Fast-Optimal FUSES the
resulting multiple paths at their crossing points (Alg. 3/4/7). This
implementation instead keeps ONE persistent RRT*-style tree (choose-parent +
rewire) and replaces the multi-path fusion with a single-path triangle-inequality
shortcut — same intent (cheap post-hoc path improvement), different mechanism;
asymptotic optimality here is carried by the RRT* rewiring, not by fusion.
Fast-Sampling and Random Steering follow the paper (Alg. 5/6, with the unbounded
resample loop capped by `steering_attempts`).
"""

from __future__ import annotations

import math
import time

import numpy as np

from navigation.core.capabilities import SamplingSpace
from navigation.core.trace import TraceRecorder
from navigation.core.types import PlanResult, PlanStats, Point

from ._sampling import (
    Tree,
    _SamplingPlanner,
    insert_best_parent,
    near_radius,
    path_length,
    rewire,
)


class FastRRT(_SamplingPlanner["SamplingSpace[Point]"]):
    @property
    def name(self) -> str:
        return "fast_rrt"

    def plan(
        self,
        space: SamplingSpace[Point],
        start: Point,
        goal: Point,
        recorder: TraceRecorder | None = None,
    ) -> PlanResult[Point]:
        t0 = time.monotonic()
        max_iter, step_size, goal_bias, goal_tol, seed = self._common_params()
        neighbor_radius = self.params.get_float("neighbor_radius")
        radius_mode = self.params.get_string("radius_mode")
        rgg_gamma = self.params.get_float("rgg_gamma")
        reached_radius = self.params.get_float("reached_radius")
        steering_attempts = self.params.get_int("steering_attempts")
        rng = np.random.default_rng(seed)
        tree = Tree(start)
        best_path: list[Point] = []
        best_cost = float("inf")
        iterations = 0
        for _ in range(max_iter):
            iterations += 1
            q_rand = self._fast_sample(
                space, tree, goal, goal_bias, reached_radius, steering_attempts, rng
            )
            if recorder is not None:
                recorder.sample_drawn(q_rand)
            near_idx = tree.nearest(q_rand, space)
            q_near = tree.points[near_idx]
            q_new = self._random_steer(space, q_near, q_rand, step_size, steering_attempts, rng)
            if q_new is None:
                continue

            # shrinking 모드는 트리 크기 n 에 따라 근방 반경을 줄인다 (Karaman & Frazzoli
            # 2011); fixed(기본)는 상수 neighbor_radius 라 기존 동작을 보존한다.
            radius = near_radius(radius_mode, neighbor_radius, rgg_gamma, len(tree))
            neighborhood = tree.near(q_new, radius, space)
            new_idx = insert_best_parent(space, tree, q_new, near_idx, neighborhood, recorder)
            rewire(space, tree, new_idx, q_new, neighborhood, recorder)

            if space.distance(q_new, goal) <= goal_tol and space.is_motion_valid(q_new, goal):
                # goal is appended to the extracted path, never inserted into the
                # tree, so it never becomes a nearest/near/rewire candidate.
                raw_path = tree.path_to(new_idx) + [goal]
                pruned = self._shortcut(space, raw_path)
                pruned_cost = path_length(space, pruned)
                if pruned_cost < best_cost:
                    best_cost = pruned_cost
                    best_path = pruned
                    if recorder is not None:
                        recorder.path_found(best_path)

        runtime = time.monotonic() - t0
        success = len(best_path) > 0
        cost = best_cost if success else 0.0
        self._finish(
            recorder, success, cost, len(tree) - 1, iterations, len(tree), iterations, runtime
        )
        stats = PlanStats(
            expanded_nodes=len(tree) - 1,
            samples=iterations,
            iterations=iterations,
            tree_size=len(tree),
        )
        return PlanResult(success, best_path, cost, stats)

    def _fast_sample(
        self,
        space: SamplingSpace[Point],
        tree: Tree,
        goal: Point,
        goal_bias: float,
        reached_radius: float,
        steering_attempts: int,
        rng: np.random.Generator,
    ) -> Point:
        # Fast-Sampling (Wu et al. 2021): a goal-biased draw is always accepted so
        # the tree can still reach the goal; a free draw is rejected while inside any
        # node's reached_radius (bounded by steering_attempts) to concentrate on
        # unreached space and cut search-time variance.
        if rng.random() < goal_bias:
            return goal
        q = space.sample()
        for _ in range(steering_attempts):
            if all(space.distance(q, node) > reached_radius for node in tree.points):
                break
            q = space.sample()
        return q

    def _random_steer(
        self,
        space: SamplingSpace[Point],
        q_near: Point,
        q_rand: Point,
        step_size: float,
        steering_attempts: int,
        rng: np.random.Generator,
    ) -> Point | None:
        q_new = space.steer(q_near, q_rand, step_size)
        if space.is_motion_valid(q_near, q_new):
            return q_new
        # Random Steering: probe random directions to escape a blocked extension.
        for _ in range(steering_attempts):
            angle = float(rng.uniform(0.0, 2.0 * math.pi))
            cand = (
                q_near[0] + step_size * math.cos(angle),
                q_near[1] + step_size * math.sin(angle),
            )
            if space.is_motion_valid(q_near, cand):
                return cand
        return None

    def _shortcut(self, space: SamplingSpace[Point], path: list[Point]) -> list[Point]:
        # Fast-Optimal: single greedy pass that jumps from each kept waypoint to
        # the farthest later waypoint reachable by a valid straight segment. Kept
        # byte-for-byte equivalent to the C++ FastRrtPlanner::shortcut_prune so the
        # cross-language benchmark compares the same shortcut on occluded paths.
        if len(path) <= 2:
            return path
        out = [path[0]]
        i = 0
        while i < len(path) - 1:
            j = len(path) - 1
            while j > i + 1 and not space.is_motion_valid(path[i], path[j]):
                j -= 1
            out.append(path[j])
            i = j
        return out
